### ISSUE TITLE
Wrong checksum detection when scanning single file instead of folder

### ISSUE THREAD

**Issue by ner00** [libretro/RetroArch#10704]
I have a PSP iso which I scan individually and RetroArch catalogues it wrongly and attributes the wrong checksum to it. Doing teh same scan t the whole folder correctly identifies the version and checksum; this is clearly a bug.

The ISO is undoubtedly this one:
```
Label: Tekken 6 (USA) (En,Fr,De,Es,It,Ru)
CRC32: AE9ABFB0
MD5: 740F4FB82775D4E8C1843FA9777F5389
SHA-1: BA1C98D3B30F296C4FE03A68BB7CE51738E069D6
Serial: ULUS-10466
```

**Scanning ISO individually (wrong detection):**
```
"path": "I:\\RetroArch\\roms\\Sony - PlayStation Portable\\Tekken 6 (USA).iso",
"label": "Tekken 6 (USA) (En,Fr,De,Es,It,Ru) (2014-05-20) (PSP) (PSN)",
"crc32": "B731B730|crc",
"db_name": "Sony - PlayStation Portable (PSN).lpl"
```

**Scanning the whole folder, same ISO (correct detection):**
```
"path": "I:\\RetroArch\\roms\\Sony - PlayStation Portable\\Tekken 6 (USA).iso",
"label": "Tekken 6 (USA) (En,Fr,De,Es,It,Ru)",
"crc32": "AE9ABFB0|crc",
"db_name": "Sony - PlayStation Portable.lpl"
```

I always assumed it calculated a checksum, but it is clearly not doing that in this case when using the "Scan File" option. Curiously the game is correct, but the checksum is clearly not. It doesn't seem to fallback on the ISO filename because something like `foobar.iso` will still get catalogued wrongly but consistently the correct game. I had never tried PSP games in RA so I'm not sure if this bug was present before the current version.

RetroArch: 1.8.6/d3a55ef732

**Comment by i30817** [comment]
funny thing on a bunch of console targets (not sure about psp), the non-manual scanner is using serials. For 'performance'.

So if the rdb is non-deterministic about which entry is returned first, this could happen easily. Here is another manifestation of the same bug:

https://github.com/libretro/RetroArch/issues/7455

check if information about the iso in the RA playlist entry contains duplicate entries.

**Comment by Sanaki** [comment]
PSP is serial-based as well, yes.

**Comment by ner00** [comment]
> funny thing on a bunch of console targets (not sure about psp), the non-manual scanner is using serials. For 'performance'.

Well, that explains it... my ROM is being inaccurately matched because it has the same serial of multiple releases:

**Sony - PlayStation Portable (PSN).dat**
```
name "Tekken 6 (USA) (En,Fr,De,Es,It,Ru) (2014-05-20) (PSP) (PSN)"
description "Tekken 6 (USA) (En,Fr,De,Es,It,Ru) (2014-05-20) (PSP) (PSN)"
origin "USA"
serial "ULUS-10466"
rom ( name "Tekken 6 (USA) (En,Fr,De,Es,It,Ru) (2014-05-20) (PSP) (PSN).iso" size 855638016 crc B731B730 md5 BF5D645CF5D3679CC183792C6DB388D4 sha1 A019B35F435BFB58EA63463AC4707F0EA9956DC7 serial "ULUS-10466" )
```

**Sony - PlayStation Portable.dat**
```
name "Tekken 6 (USA) (En,Fr,De,Es,It,Ru)"
description "Tekken 6 (USA) (En,Fr,De,Es,It,Ru)"
origin "USA"
serial "ULUS-10466"
rom ( name "Tekken 6 (USA) (En,Fr,De,Es,It,Ru).iso" size 855638016 crc AE9ABFB0 md5 740F4FB82775D4E8C1843FA9777F5389 sha1 BA1C98D3B30F296C4FE03A68BB7CE51738E069D6 serial "ULUS-10466" )
```

Not sure why the folder scan is any different, maybe it goes directly to the database file of the same name as the folder; anyway, that's beside the point, more importantly is that this has been reported for over a year and serial identification as primary ID method is still going for a few of the platforms, the bug is indeed minor all the more reason to get it over with and close the issue. It can, as mentioned in https://github.com/libretro/RetroArch/issues/7455, be used for secondary stuff but certainly not for unique identification as is the case with any rom/image.

**Comment by i30817** [comment]
I'm going to dash your expectations here: serial identification has been *expanded* since then. Gotta go fast.

**Comment by i30817** [comment]
BTW multiple releases from different dumping groups could also have the same checksum, so even the crc check is not enough to remove all kinds of duplicates; unless libretro-database does something sane and remove duplicate hashes and only allow '1 canonical name' for each hash. 

Something like 'redump > nointro > tosec > whatever' would have to be implemented to filter out those cheeky duplicates.

**Comment by ner00** [comment]
> unless libretro-database does something sane and remove duplicate hashes and only allow '1 canonical name' for each hash

That about sums it up.
That metadata is also associated with assets (thumbs, etc), you would need to have also duplicate assets for each different rom name, and boy does libretro-thumbnais have dupes...

**Comment by i30817** [comment]
Libretro-thumbnails could be using symlinks right now if someone that could do a github bot stepped up and did some post-merge bottage.

It would like this:
1. committer commits the 'm3u name' or single game cd name `Gamename (Region).png` (extension optional). Could also commit a 'real' name for cases where you don't want the bot to do anything.
2. repository admin merges.
3. boot notices and fetches the .png's added. Checks that after removing all the 'groups' `(text)` of dumping groups .dat files for that subrepo (they are for consoles) find those that match the png files committed. For instance:

redump.dat `Xenogears (Disc 1) (USA).cue` matches `Xenogears (USA).png`, so the bot creates a symlink to the png file with that name; but if the Xenogears had a second super cool revision with a different cover and the committer decided to add `Xenogears (USA) (Rev 2).png` (for both disc 1 and 2)  the bot would first sort both files so the longest filenames come first (with more metadata), so it would naturally create those symlinks first, and refuse to create symlinks that already exist.

Commits which remove stuff the bot does nothing.

**Comment by Sanaki** [comment]
With the expansion of chd support, there's the option of using chd's sha1 as a secondary verification for distinction between variants without any noteworthy hit to speed. That only works for systems where chd is supported though. PSP will undoubtedly join that list soonish, and I'm sure PS2 will eventually as well, but it isn't an option for large cartridges where you'd need a fast hash, such as DS. In theory you could add an option to use crc32 as a backup distinction when multiple entries share a serial. But frankly, serial scanning is a good thing in general.

**Comment by i30817** [comment]
I disagree. Especially on systems that didn't even need serial scanning. Like say, the snes.

**Comment by Sanaki** [comment]
The point of serial scanning is to simplify the process of scanning for an average user who just wants to hit scan and have their game show up. Retroarch's scanner is not and will never be a replacement for a rom manager. Treating it like one is just a mistake. If you want to use the auto-scanner and have it be able to differentiate between versions, the best bet is to still use serial scanning, with a fallback secondary verification when multiple entries match. If you use an actual rom manager, using the auto-scanner is only doing you a disservice in the first place.

**Comment by i30817** [comment]
Who wants the RA as a rom manager. I don't. I want the games to appear with the right metadata. I want metadata not to clash because there are multiply entries. I want useful metadata that can't be placed on the rdbs because it would be useless with a faulty lookup method. I want hacks showing their real names instead 3 version of the original games as entries because the original game and 2 hacks.

**Comment by zoltanvb** [comment]
The original problem (different identification method is used with folder scan and file scan) should not exist now, as I checked quickly.

About using crc as a secondary identification in addition to serial, it was introduced in #15623.

**Comment by i30817** [comment]
I've quit trying to convince devs, but I still believe that RetroArch would benefit from slightly (only slightly) simplifying the manual scanner (as I've opened feature requests about) and just simply removing altogether the automatic scanner, crcs or serials. 

Yes, it gets rid of any possibility of RA libretro (not emulator core) custom hacks. Those never existed and will never exist apparently. Even the much hyped explore tab never worked correctly. Because it can't accommodate name only playlists and because the info is simply lacking.

The 'benefit' of accommodating fools that don't even have the brainpower to place roms for the same system in the same directory tree is already negated by all the other settings.

RetroArch is full of custom and global options that do nothing, do something poorly or give users a nice shotgun to sabotage some other feature. The default options of the manual scanner sabotage thumbnails download if you don't name the scanned folders exactly correct ffs. The information tab will give you absolutely fictional matches for a single rom, or call a hack the original, etc. Bitrotten areas of code no one can feel ownership to change or the political will to cut off a malfunctioning 'feature' are plenty, and it's not just around databases. Why is a _progress bar_ affecting the **speed** of  savestates and why is dolphin upstream savestates 3 times faster as baseline (answer: they don't try to use a progress bar and use the mandatory thread -if you don't want to be many more times slower- for actual emulation too).  The ironic thing is that I'm sure if by some miracle a dev decided to fix this they'd be forced to 'make it a nondefault option'. Not that using the dolphin core instead of the upstream emulator is even a good idea, considering all the crashes and tens of thousands of commits behind.

Users calling RetroArch bloated and then start using another libretro frontend is the most ironic thing of it all.
